### DISCUSSION TITLE
Suport M:N Bidirectional relationship with MongoDB

### DISCUSSION THREAD

**Discussion by bahdcoder** [mikro-orm/mikro-orm#1170] [Q&A]
**Is your feature request related to a problem? Please describe.**
At the moment, only one side of a m:n relationship in MongoDB can query data such as count and find. Here's an example:

Consider a Tag entity:

```javascript
{
       "name": "Tag",
       "tableName": "tags",
       "collection": "tags",
       "properties": {
          "_id": {
             "type": "ObjectId",
             "primary": true
          },
          "posts": {
             "name": "posts",
             "type": "Post",
             "primary": false,
             "reference": "m:n",
             "owner": true,
             "inversedBy": "tags",
             "mappedBy": "tags"
          }
       }
    }
```
And a Post entity:

```javascript
 {
       "name": "Post",
       "tableName": "posts",
       "collection": "posts",
       "properties": {
          "_id": {
             "type": "ObjectId",
             "primary": true
          },
          "tags": {
             "name": "tags",
             "type": "Tag",
             "primary": false,
             "reference": "m:n",
             "owner": true,
             "mappedBy": "posts",
             "inversedBy": "posts"
          },
       }
    }
```

Only the tags collection would have a reference to the posts, and this limits greatly the kind of queries that can be run on posts about tags.

For example, how do we find a list of posts where has tag `tag1` ? Now how do we add additional conditions to this query ? And maybe limit the result of posts we get ? 

**Describe the solution you'd like**
It would be great to be able to query both sides of the relationship, which means saving the references on both sides, and being able to query both sides when need be.

**Describe alternatives you've considered**
To solve the above example I gave, I had to use a `$lookup` aggregation. 

**Additional context**
Add any other context about the feature request here.

**Comment by dvlalex** [comment]
I think @B4nan already mentioned this in an old issue, but what you want is actually not possbile because of mongo's limitations.

While having references on both sides would be great, working with mutable arrays in mongo is not recommended, and there's no other way to reference M:N documents. Using an aggregation pipeline is the right way to deal with these kinds of queries.

**Comment by bahdcoder** [comment]
Thanks for sharing @dvlalex . Is there any limitation to the framework handling this query behind the scenes ? That way entities and collections are also correctly initialised after the query?

If it can be done, then the framework could handle it to make life very easy for anyone wanting this functionality. 

We can't build something as basic as a blog with Posts and Tags without this feature.

**Comment by B4nan** [comment]
Having the relations on both ends means maintaining both ends. It would be literally 2 owning sides, not a bidirectional relation. You can already do that, obviously you will need to maintain both sides in sync yourself. 

No plans to do this at the ORM level, it would mean a lot of querying in the background, you are basically asking for denormalization of the schema (at this project goes pretty much the opposite way, there is not a single place where we would denormalize the schema). If something, this would have to work the same way as in SQL drivers, so via pivot table. But as there are no joins, it would have some consequences/differences to the way it works now (e.g. loading the relation would do 2 queries instead of 1, and more importantly, it would be impossible to query the parent/owning entity by the collection).

Using `$lookup` is the way to go. Or as I mentioned above, you can use 2 owning sides to have the array on both ends. You will need to handle changes on both ends - if you remove entity, you will need to look up all connected entities on both ends, remove the item from those, and similar changes would be required for adding to such collection. Personally I would strongly advice to rather use normalized schema

Btw looking at your entity definitions, using both `inversedBy` and `mappedBy` on one property have no meaning. It is literally the same thing, its just about the side that is defining the relation (the side that is using `inversedBy` is the owning side, and value points to the inverse side). Also I don't understand why are you setting both relations to `owner: true`, only one side of bidirectional relation is owner (the one that uses `inversedBy` option). To define two owning sides, just drop all of `owner/inversedBy/mappedBy` - all of that is meant only for bidirectional relations, omitting those means the relation is owner.

**Comment by bahdcoder** [comment]
@B4nan thanks very much for your clear explanation. 

Using `owner: true` was my test attempt of having two owning sides. Thankfully you've covered how to define two owning sides in your explanation. 

Thanks @B4nan and @dvlalex. 
Hopefully this helps someone in future too.
